### PR TITLE
Add sensor parsing TypeScript code

### LINKED ISSUE #1289: Add sensor parsing TypeScript code
We should use existing server running on Node to parse TypeScript files.
Parsing errors should be reported.

To parse TypeScript we should use `@typescript-eslint/parser`.

### PR DESCRIPTION
**Pull request by saberduck**
Fixes #1289

### CODE REVIEW

File: sonar-javascript-plugin/src/main/java/org/sonar/plugins/javascript/eslint/EslintBridgeServer.java [hunk 1]
```diff
@@ -20,24 +20,89 @@
...
 
 @ScannerSide
 @SonarLintSide(lifespan = MULTIPLE_ANALYSES)
 public interface EslintBridgeServer extends Startable {
 
-  void startServerLazily(SensorContext context) throws IOException, ServerAlreadyFailedException, NodeCommandException;
+  void startServerLazily(SensorContext context) throws IOException;
 
-  String call(String request) throws IOException;
+  AnalysisResponseIssue[] call(AnalysisRequest request) throws IOException;
```
Comment (vilchik-elena) [inline sonar-javascript-plugin/src/main/java/org/sonar/plugins/javascript/eslint/EslintBridgeServer.java]: why `call`? should we rename it to `analyzeJavaScript `

File: its/plugin/projects/tsproject/tsconfig.json [hunk 2]
```diff
@@ -0,0 +1,6 @@
+{
```
Comment (vilchik-elena) [inline its/plugin/projects/tsproject/tsconfig.json]: should we also have a test with multiple tsconfigs in the project?

File: sonar-javascript-plugin/src/main/java/org/sonar/plugins/javascript/eslint/TypeScriptSensor.java [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+    } catch (IOException e) {
+      LOG.error("Failed to get response while analyzing " + file, e);
+    }
+  }
+
+  private List<String> tsConfigs(SensorContext context) {
+    if (tsconfigs == null) {
+      FileSystem fs = context.fileSystem();
+      tsconfigs = new ArrayList<>();
+      fs.inputFiles(fs.predicates().hasRelativePath("tsconfig.json")).forEach(file -> tsconfigs.add(file.absolutePath()));
```
Comment (vilchik-elena) [inline sonar-javascript-plugin/src/main/java/org/sonar/plugins/javascript/eslint/TypeScriptSensor.java]: I think we discussed that we should rely on physical file system to discover tsconfigs, as we might not find them when they are excluded from analysis for some reason

File: sonar-javascript-plugin/src/main/java/org/sonar/plugins/javascript/eslint/EslintIssue.java [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+import org.sonar.api.batch.sensor.SensorContext;
+import org.sonar.api.batch.sensor.issue.NewIssue;
+import org.sonar.api.batch.sensor.issue.NewIssueLocation;
+import org.sonar.api.rule.RuleKey;
+import org.sonar.plugins.javascript.JavaScriptChecks;
+
+import static org.sonar.plugins.javascript.eslint.EslintBridgeServer.AnalysisResponseIssue;
+import static org.sonar.plugins.javascript.eslint.EslintBridgeServer.IssueLocation;
+
+class EslintIssue {
```
Comment (vilchik-elena) [inline sonar-javascript-plugin/src/main/java/org/sonar/plugins/javascript/eslint/EslintIssue.java]: let's call it `EslintBasedIssue`

File: sonar-javascript-plugin/src/test/java/org/sonar/plugins/javascript/eslint/EslintBasedRulesSensorTest.java [hunk 5]
```diff
@@ -326,7 +314,7 @@ public void handle_missing_node() throws Exception {
     EslintBasedRulesSensor eslintBasedRulesSensor = new EslintBasedRulesSensor(checkFactory(ESLINT_BASED_RULE), eslintBridgeServerMock, analysisWarnings);
     eslintBasedRulesSensor.execute(context);
     assertThat(logTester.logs(LoggerLevel.ERROR)).contains("Exception Message");
-    verify(analysisWarnings).addUnique("Some JavaScript rules were not executed. Exception Message");
+    verify(analysisWarnings).addUnique("Some eslint-based rules were not executed. Exception Message");
```
Comment (vilchik-elena) [inline sonar-javascript-plugin/src/test/java/org/sonar/plugins/javascript/eslint/EslintBasedRulesSensorTest.java]: I'm not sure it's correct to say this way, if no node, all (not some) eslint-based rules will not be executed

File: sonar-javascript-plugin/src/test/java/org/sonar/plugins/javascript/eslint/EslintBridgeServerImplTest.java [hunk 6]
```diff
@@ -125,7 +128,27 @@ public void should_get_answer_from_server() throws Exception {
...
+    assertThat(eslintBridgeServer.call(request)).isEqualTo(new EslintBridgeServer.AnalysisResponseIssue[0]);
+  }
+
+  @Test
+  public void should_get_answer_from_server_for_ts_request() throws Exception {
+    eslintBridgeServer = createEslintBridgeServer(START_SERVER_SCRIPT);
+    eslintBridgeServer.deploy();
+    eslintBridgeServer.startServer(context);
+
+    DefaultInputFile inputFile = TestInputFileBuilder.create("foo", "foo.js")
```
Comment (vilchik-elena) [inline sonar-javascript-plugin/src/test/java/org/sonar/plugins/javascript/eslint/EslintBridgeServerImplTest.java]: `foo.ts` may be better here?

File: sonar-javascript-plugin/src/test/java/org/sonar/plugins/javascript/eslint/EslintBridgeServerImplTest.java [hunk 7]
```diff
@@ -125,7 +128,27 @@ public void should_get_answer_from_server() throws Exception {
...
+      .setContents("alert('Fly, you fools!')")
+      .build();
+    DefaultInputFile tsConfig = TestInputFileBuilder.create("foo", "tsconfig.json")
+      .setContents("{\"compilerOptions\": {\"target\": \"es6\", \"allowJs\": true }}")
+      .build();
+    EslintBridgeServer.TypeScriptAnalysisRequest request = new EslintBridgeServer.TypeScriptAnalysisRequest(inputFile, new EslintBridgeServer.Rule[0], Collections.singletonList(tsConfig.absolutePath()));
```
Comment (vilchik-elena) [inline sonar-javascript-plugin/src/test/java/org/sonar/plugins/javascript/eslint/EslintBridgeServerImplTest.java]: line is too long

File: sonar-javascript-plugin/src/test/java/org/sonar/plugins/javascript/eslint/EslintBridgeServerImplTest.java [hunk 8]
```diff
@@ -186,6 +209,26 @@ public void should_throw_special_exception_when_failed_already() throws Exceptio
...
+  public void should_not_explode_if_bad_json_response() throws Exception {
+    eslintBridgeServer = createEslintBridgeServer("badResponse.js");
+    eslintBridgeServer.deploy();
+    eslintBridgeServer.startServerLazily(context);
+
+    DefaultInputFile inputFile = TestInputFileBuilder.create("foo", "foo.js")
+      .setContents("alert('Fly, you fools!')")
+      .build();
+    EslintBridgeServer.AnalysisRequest request = new EslintBridgeServer.AnalysisRequest(inputFile, new EslintBridgeServer.Rule[0]);
+    eslintBridgeServer.call(request);
```
Comment (vilchik-elena) [inline sonar-javascript-plugin/src/test/java/org/sonar/plugins/javascript/eslint/EslintBridgeServerImplTest.java]: should we check return of this call?

File: eslint-bridge/src/parser.ts [hunk 9]
```diff
@@ -68,6 +68,27 @@ export function parseSourceFile(fileContent: string, fileUri: string): SourceCod
...
+  fileContent: string,
+  fileUri: string,
+  tsConfigs: string[],
+): SourceCode | undefined {
+  try {
+    // we load the typescript parser dynamically, so we don't need typescript dependency when analysing pure JS project
+    const tsParser = require("@typescript-eslint/parser");
+    const result = tsParser.parseForESLint(fileContent, {
+      ...PARSER_CONFIG_MODULE,
+      filePath: fileUri,
```
[the remark this answers is not in the document]
Comment (saberduck) [reply eslint-bridge/src/parser.ts]: name is misleading, we don't use uri, we use path, even if it is deprecated. I will rename the variable to make it clear

File: sonar-javascript-plugin/src/main/java/org/sonar/plugins/javascript/JavaScriptExclusionsFileFilter.java [hunk 10]
```diff
@@ -39,10 +39,6 @@ public JavaScriptExclusionsFileFilter(Configuration configuration) {
 
   @Override
   public boolean accept(InputFile inputFile) {
-    if (!JavaScriptLanguage.KEY.equals(inputFile.language())) {
-      return true;
-    }
-
     String[] excludedPatterns = this.configuration.getStringArray(JavaScriptPlugin.JS_EXCLUSIONS_KEY);
```
[the remark this answers is not in the document]
Comment (saberduck) [reply sonar-javascript-plugin/src/main/java/org/sonar/plugins/javascript/JavaScriptExclusionsFileFilter.java]: done https://github.com/SonarSource/SonarJS/issues/1311

File: sonar-javascript-plugin/src/main/java/org/sonar/plugins/javascript/eslint/TypeScriptSensor.java (hunk 3 above)
Comment (saberduck) [reply sonar-javascript-plugin/src/main/java/org/sonar/plugins/javascript/eslint/TypeScriptSensor.java]: let's do that in another PR

File: eslint-bridge/src/server.ts [hunk 11]
```diff
@@ -42,6 +42,7 @@ export function start(port = 0): Promise<Server> {
     app.post("/analyze-ts", (request: express.Request, response: express.Response) => {
       const parsedRequest = request.body as TypeScriptAnalysisInput;
       const analysisResponse = analyzeTypeScript(parsedRequest);
+      console.log(JSON.stringify(request.body));
```
Comment (vilchik-elena) [inline eslint-bridge/src/server.ts]: do we need this log?

Comment (saberduck) [reply eslint-bridge/src/server.ts]: oups :)

**Pull request merged**